### PR TITLE
11.0 mig product dimension

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please include #339 and squash a bit commit history by logical set/author.

**Review by leemannd — CHANGES_REQUESTED** [review]
They are steps missing from a migration 9.0 -> 10.0 
https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0

File: product_dimension/__init__.py
Comment (leemannd) [inline product_dimension/__init__.py]: You can remove copyright from `init` files https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

File: product_dimension/__manifest__.py
Comment (leemannd) [inline product_dimension/__manifest__.py]: Copyright symbol can be replaced s/©/Copyright/

File: product_dimension/models/product.py
Comment (leemannd) [inline product_dimension/models/product.py]: All `openerp` import should be replaced by `odoo`

**Review by leemannd — CHANGES_REQUESTED** [review]

**Comment by leemannd** [comment]
Sidenote: there is an ongoing migration for V11. But it is not going forward => https://github.com/OCA/product-attribute/pull/306
There is also a migration for V10.0 that we should take care of. With a discussion about the views https://github.com/OCA/product-attribute/pull/195/files

**Comment by ghost** [comment]
@leemannd thanks for your input
I removed dependency to stock and added suggested improvement aswell as german translation

File: product_dimension/__init__.py
Comment (asaunier) [inline product_dimension/__init__.py]: Minor point but this line is no longer required in v11 (Python 3), see https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0#tasks-to-do-in-the-migration
(ditto for the other py files)

File: product_dimension/README.rst
Comment (asaunier) [inline product_dimension/README.rst]: Using the full URL https://github.com/adhoc-dev/odoo-addons would make the link clickable.

File: product_dimension/README.rst
Comment (asaunier) [inline product_dimension/README.rst]: You can add your name here :)

File: product_dimension/i18n/de.po
Comment (asaunier) [inline product_dimension/i18n/de.po:1]: Nice!
But aren't [the other languages available in v9](https://github.com/OCA/product-attribute/tree/9.0/product_dimension/i18n) missing?
Isn't a "OCA Transbot updated translations from Transifex" commit missing, providing the translations done in Transifex?
See https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-transifex

Comment (ghost) [reply product_dimension/i18n/de.po:1]: good point: I created migration based on 10.0 branch and there the i18n folder is missing.. 
https://github.com/OCA/product-attribute/tree/10.0/product_dimension

how can I add the missing files?

**Comment by ghost** [comment]
@asaunier thanks for the review. I've added the changes.. the i18n files are still missing.. any idea how they can be added so the git history is intact?

**Comment by asaunier** [comment]
@BT-kaberer Thanks for the changes!

It's indeed a bit weird that the i18n files have not been integrated in branch 10.0 !? I don't know if there was a reason, or maybe they were simply forgotten?

To get them back you could for instance try to create a new dummy migration branch from branch 9.0 and run the ``git format-patch`` as described in https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-10.0#technical-method-to-migrate-a-module-from-90-to-100-branch
Then you squash the "OCA Transbot" commits and cherry-pick the resulting commit to your existing branch (you might have a conflict with your ``de.po`` file).

Alternately it is maybe acceptable to copy-paste the PO files by hand in your branch and commit them yourself, since anyway [it is recommended to squash the i18n commits](https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-transifex) when migrating a module, thus not preserving the translations history.

File: product_dimension/__manifest__.py
Comment (ghost) [inline product_dimension/__manifest__.py]: as we have `product_dimension/static/description/icon.png` then maybe we should add it to images?
```
    'images': [
        'static/description/icon.png',
    ],
```

File: product_dimension/models/product.py
Comment (ghost) [inline product_dimension/models/product.py]: can be in one row )

File: product_dimension/__manifest__.py
Comment (ghost) [inline product_dimension/__manifest__.py]: `Camptocamp SA`

File: product_dimension/models/product.py
Comment (ghost) [inline product_dimension/models/product.py]: If we check if all required values are present
```
if (self.length and self.height and self.width and self.dimensional_uom_id):
```
and put `volume` calculation under this `if` then we should increase code coverage.

Same for `'product.product'`

**Review by leemannd — APPROVED** [review]

File: product_dimension/models/product.py
Comment (asaunier) [inline product_dimension/models/product.py]: With this change, the function will always return ``False``. Is it really the wanted behaviour?
Wasn't the function already OK?
Ditto for the function in ``ProductTemplate``.

**Review by asaunier — COMMENTED** [review]
A question about your changes in the product model.
Thanks for adding the translations back!

File: product_dimension/README.rst
Comment (simahawk) [inline product_dimension/README.rst]: update readme according to https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: product_dimension/README.rst
Comment (simahawk) [inline product_dimension/README.rst]: .png

File: product_dimension/models/product.py
Comment (simahawk) [inline product_dimension/models/product.py]: `domain=lambda self: self._get_dimension_uom_domain(),`

File: product_dimension/models/product.py
Comment (simahawk) [inline product_dimension/models/product.py]: these 2 methods are repeated into prod template: move them to simple functions and call them where needed

**Review by simahawk — CHANGES_REQUESTED** [review]

File: product_dimension/models/product.py
Comment (ghost) [reply product_dimension/models/product.py]: OK correct. I changed to previous behavior. Function was re-written to increase code coverage

File: product_dimension/models/product.py
Comment (ghost) [reply product_dimension/models/product.py]: I added your improvements except this one: 
1) these functions are already simple functions. 
one could, if it makes sense, encapsulate further and move the function on product.product and call it from product.template

2) the PR has already been approved by @leemannd

Comment (leemannd) [reply product_dimension/models/product.py]: @simahawk @BT-kaberer It'm not really in favour of too much refactoring nor improvements during migrations.
But in this one seems legit. It could be great if you take the time to consider @simahawk 's improvement.

Comment (nomadic-coding) [reply product_dimension/models/product.py]: @simahawk @leemannd 
done

Comment (ghost) [reply product_dimension/models/product.py]: @leemannd @simahawk 
done

File: product_dimension/README.rst
Comment (simahawk) [inline product_dimension/README.rst]: you can drop these sections if they don't provide additional info.

File: product_dimension/models/product.py
Comment (simahawk) [inline product_dimension/models/product.py]: `if volume is False:`

File: product_dimension/models/product.py
Comment (simahawk) [inline product_dimension/models/product.py]: `if volume is False:` is more explicit.
Also, since this is an onchange, when volume is not valued don't you want to reset the value?
If parameters do not match then the volume should be updated as well. No?

File: product_dimension/README.rst
Comment (simahawk) [inline product_dimension/README.rst]: drop empty sections

**Review by simahawk — CHANGES_REQUESTED** [review]

File: product_dimension/models/product.py
Comment (leemannd) [reply product_dimension/models/product.py]: Thanks

File: product_dimension/models/product.py
Comment (ghost) [reply product_dimension/models/product.py]: yes you are right. i changed it. It was just so the function behaved like before migration. but now make more sense to me

**Review by asaunier — APPROVED** [review]

**Review by simahawk — APPROVED** [review]

**Comment by leemannd** [comment]
@oleksandrpaziuk Could you update your review?

**Review by ghost — APPROVED** [review]
LGTM!
But newline required in the end of *.po files

**Review by ivantodorovich — APPROVED** [review]
LGTM

**Review by jjscarafia — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Can you please squash your migration commits together before merging?

**Review by florian-dacosta — APPROVED** [review]
Seems good.
One question though, wouldn't it be better to put the volume field as a computed stored field?
This module is based only on the onchange mechanism, which is a bit weak for this kind of data no?
For instance, if I always update my products by csv file import, if I change the dimension, the volume won't be recomputed right?

**Comment by ivantodorovich** [comment]
@florian-dacosta the thing is that if we use a calculated field, you then loose the posibility to edit the volume. Maybe we can achieve the same with and inverse function and a hidden field.

I agree that using onchange for this is not ideal. I've seen some weird behaviours with this module in previous versions, sadly I don't remember them to replicate

**Comment by LoisRForgeFlow** [comment]
As a migration effort, isn't this ready to merge? Any extra discussion/improvement belong to a separate issue/PR at this point I think.

**Comment by asaunier** [comment]
@lreficent +1
6 approvals and no request for change it seems. Is there still something missing before the PR is merged?

**Review by rvalyi — APPROVED** [review]

**Comment by rvalyi** [comment]
dear contributors, so after approving this PR myself too yesterday I went with the merge. But I let you know that today I made a new PR to migrate product_dimension further to v12 here
https://github.com/OCA/product-attribute/pull/485
So your review is welcome...

**Pull request merged**
